share/system/cgroup_linux.go

```diff
@@ -332,76 +332,147 @@ func getCgroupPath_cgroup_v2(pid int) (string, error) {
 
 }
 
+// GetContainerCgroupPath - Gets a PID's cgroup path for a subsystem
+// PIDs can share containers so we can derive the pid's cgroup path by parsing proc/<pid>/cgroup
 func (s *SystemTools) GetContainerCgroupPath(pid int, subsystem string) (string, error) {
-	/*
-		mnt, err := FindCgroupMountpoint(subsystem)
-		if err != nil {
-			return "", nil
-		}
-	*/
+	subsystemPath := ""
+	mpath, err  := s.getCgroupMetricsPath(pid, subsystem)
+	if err != nil {
+		// Just return the fallback, callers don't care about the error!
+		return mpath, nil
+	}
 
-	// Alternative: it might not be necessary to obtain cgroup path as before
-	var path string
+	// Join to the system's path for host's cgroup
+	subsystemPath = s.JoinToCgroupPath(mpath, subsystem)
+
+	return subsystemPath, nil
+}
+
+// JoinToCgroupPath - Join to the system's path for host's cgroup
+// For cgroup v1, the subsystems are directories in the root and then the namespaces are within them
+// For cgroup v2, a flat directory structure is used and the namespace's directory now holds all the subsystems files
+func (s *SystemTools) JoinToCgroupPath(path string, subsystem string) string {
+	subsystemPath := ""
 	switch s.cgroupVersion {
+	case cgroup_v2:
+		subsystemPath = filepath.Join(s.cgroupDir, path)
+	// unsupported - Found issues in ubuntu 18.04 and missing subsystem in the unified directories
+	//case cgroup_v2_hybrid:
+	//	subsystemPath = filepath.Join(s.cgroupDir, "/unified", mpath)
 	case cgroup_v1:
-		// It is a well-known path: /proc/<pid>/root/sys/fs/cgroup/<subsystem>
-		if pid == 0 { // self
-			path = filepath.Join("/sys/fs/cgroup", subsystem)
-		} else {
-			path = filepath.Join(s.procDir, strconv.Itoa(pid), "root/sys/fs/cgroup", subsystem)
+		fallthrough
+	default:
+		subsystemPath = filepath.Join(s.cgroupDir, subsystem, path)
+	}
+	return subsystemPath
+}
+
+
+
+// getStatsPathFromCgroupFile - Opens the proc/<pid>/cgroup file and parses it for the subsystem and paths
+func getStatsPathFromCgroupFile(f io.Reader, subsystem string) (string, error) {
+	subsystemMap := make(map[string]string, 0)
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+
+		tokens := strings.Split(scanner.Text(), ":")
+
+		if len(tokens) > 2 {
+			// https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/sec-default_cgroup_hierarchies
+			// https://docs.fedoraproject.org/en-US/quick-docs/understanding-and-administering-systemd/
+			// cgroup path is defined by <index>:<subsystem(s)>:<path>
+			// example: 2:cpuset,cpu,cpuacct,memory,net_cls,net_prio,hugetlb:/kubepods/besteffort/pod1fe19bf5-e8ef-11e8-900c-52daee5a874d/da31e536c8d61304a6d5998d163d12400a7a9a1003e1d86369e8fadb022fc17d
+
+			// The subsystem is the second item in the line
```

Comment:
Logic should be simplified here because instead of all the fallback logic, I simplified it to use a map and we just update the map with the subsystem and paths. 

The subsystem map can be cached with the container if we care enough about it.